tests/schema/tests.py

```diff
@@ -2193,28 +2194,53 @@ def test_unique_together_with_fk(self):
 
-    def test_unique_together_with_fk_with_existing_index(self):
+    def test_composed_indexes_with_fk_with_existing_index(self):
         """
-        Tests removing and adding unique_together constraints that include
+        Tests removing and adding various type of indexes that include
         a foreign key, where the foreign key is added after the model is
         created.
         """
         # Create the tables
         with connection.schema_editor() as editor:
             editor.create_model(Author)
             editor.create_model(BookWithoutAuthor)
-            new_field = ForeignKey(Author, CASCADE)
-            new_field.set_attributes_from_name('author')
-            editor.add_field(BookWithoutAuthor, new_field)
-        # Ensure the fields aren't unique to begin with
-        self.assertEqual(Book._meta.unique_together, ())
-        # Add the unique_together constraint
-        with connection.schema_editor() as editor:
+            author_field = ForeignKey(Author, CASCADE)
+            author_field.set_attributes_from_name('author')
+            editor.add_field(BookWithoutAuthor, author_field)
+
+        @contextmanager
+        def editor_with_post_cleanup(meta_field):
+            # check that the initial state of the meta field is empty
+            # e.g. no unique constraints, indexes, etc.
+            self.assertEqual(list(getattr(Book._meta, meta_field)), [])
+            with connection.schema_editor() as editor:
+                # perform actions using the editor class
+                yield editor
+                # check that the result state of the meta field is empty
+                self.assertEqual(list(getattr(Book._meta, meta_field)), [])
+                # remove the author field as we should check creating missing
+                # index each time
+                editor.remove_field(Book, author_field)
+                editor.add_field(BookWithoutAuthor, author_field)
+
+        with editor_with_post_cleanup('unique_together') as editor:
             editor.alter_unique_together(Book, [], [['author', 'title']])
-        # Alter it back
-        with connection.schema_editor() as editor:
             editor.alter_unique_together(Book, [['author', 'title']], [])
 
+        with editor_with_post_cleanup('index_together') as editor:
+            editor.alter_index_together(Book, [], [['author', 'title']])
+            editor.alter_index_together(Book, [['author', 'title']], [])
+
+        constraint = UniqueConstraint(fields=['author', 'title'], name='author_title_uniq')
+        with editor_with_post_cleanup('constraints') as editor:
+            editor.add_constraint(Book, constraint)
+            editor.remove_constraint(Book, constraint)
+
+        index = Index(fields=['author', 'title'], name='author_title_index')
+        with editor_with_post_cleanup('indexes') as editor:
+            editor.add_index(Book, index)
+            editor.remove_index(Book, index)
+
```

Comment:
Can we add test for renaming a `ForeignKey` included in the composed index?

Reply:
Should we test that actually? As stated in the ticket, for the field renaming Django adds 3 operations to the migration file: remove an index, rename the field, add the index (back). I believe it should be at least a separate ticket, that changes behavior to one operation instead of three...